4.0rc4 won't compile with external PMIx 2.1.2

I'm having hard time getting v4 to work on our SLURM cluster. I configure it with external PMIx / HWLOC / libevent libraries:

```
./configure --prefix=[...] --enable-mpi-cxx --with-pmi --with-pmix=external --with-hwloc=external --with-libevent=external --with-mxm=${HPCX_MXM_DIR} --with-hcoll=${HPCX_HCOLL_DIR} --with-ucx=${HPCX_UCX_DIR} --with-knem=${KNEM_DIR} --with-platform=contrib/platform/mellanox/optimized
```
configure works ok
```
Miscellaneous
-----------------------
CUDA support: no
HWLOC support: external
Libevent support: external
PMIx support: 2x
```
but compilation fails
```
make[2]: Entering directory `/tmp/marcink/openmpi-4.0.0rc4/opal/mca/pmix/ext2x'
  CC       mca_pmix_ext2x_la-ext2x_local.lo
  CC       mca_pmix_ext2x_la-ext2x.lo
  CC       mca_pmix_ext2x_la-ext2x_client.lo
  CC       mca_pmix_ext2x_la-ext2x_component.lo
  CC       mca_pmix_ext2x_la-ext2x_server_north.lo
  CC       mca_pmix_ext2x_la-ext2x_server_south.lo
ext2x_local.c:21:10: fatal error: ext2x.h: No such file or directory
 #include "ext2x.h"
          ^~~~~~~~~
compilation terminated.
make[2]: *** [mca_pmix_ext2x_la-ext2x_local.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
ext2x_component.c:26:10: fatal error: ext2x.h: No such file or directory
 #include "ext2x.h"
          ^~~~~~~~~
compilation terminated.
make[2]: *** [mca_pmix_ext2x_la-ext2x_component.lo] Error 1
ext2x.c:42:10: fatal error: ext2x.h: No such file or directory
 #include "ext2x.h"
          ^~~~~~~~~
compilation terminated.
ext2x_client.c:37:10: fatal error: ext2x.h: No such file or directory
 #include "ext2x.h"
          ^~~~~~~~~
compilation terminated.
make[2]: *** [mca_pmix_ext2x_la-ext2x.lo] Error 1
ext2x_server_south.c:43:10: fatal error: ext2x.h: No such file or directory
 #include "ext2x.h"
          ^~~~~~~~~
compilation terminated.
make[2]: *** [mca_pmix_ext2x_la-ext2x_client.lo] Error 1
ext2x_server_north.c:39:10: fatal error: ext2x.h: No such file or directory
 #include "ext2x.h"
          ^~~~~~~~~
compilation terminated.
```
On the other hand, if I try to use the internal PMIx, then it configures and compiles fine
```
Miscellaneous
-----------------------
CUDA support: no
HWLOC support: external
Libevent support: external
PMIx support: internal
```
but `mpirun` fails to start anything due to a PMIx problem:
```
mpirun hostname
[c11-5.cluster:53592] [[42157,0],0] ORTE_ERROR_LOG: Not found in file ess_hnp_module.c at line 321
--------------------------------------------------------------------------
It looks like orte_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during orte_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  opal_pmix_base_select failed
  --> Returned value Not found (-13) instead of ORTE_SUCCESS
--------------------------------------------------------------------------
```

I guess those are two different problems. We need to compile with the SLURM plugin, hence we need the external PMIx (and hwloc + libevent) more. But the second problem seems more fundamental. `mpirun` fails even when run not from within a SLURM allocation.

Any ideas?

Marcin

@rhc54 The patch didn't work: the PR modifies the `ext2x.h`, which indeed doesn't exist. So I copied one from `master` and then applied the patch, which worked. I then needed to re-run `aclocal/automake/autoconf` to get things working (your patches modified the configury..). Configure went fine, but when I compile I now see this:

```
make[2]: Entering directory `/tmp/marcink/openmpi-4.0.0rc4/opal/mca/crs'
  CCLD     libmca_crs.la
  GENERATE opal_crs.7
Option package-version requires an argument
Usage: ../../../config/make_manpage.pl --package-name=<package name> --package-version=<package version> --ompi-date=<ompi date> --opal-date=<opal date> --orte-date=<orte date> --input=<input file> --output=<output file> [--nocxx] [ --nofortran] [--nof08]
make[2]: *** [opal_crs.7] Error 1
```
So I guess my `autoconf` didn't work well enough. I did see this error:

```
sh: 1: config/opal_get_version.sh: not found
```
Can I fix this? Otherwise maybe you can provide a tar ball?

m

--- a/config/opal_check_pmi.m4
+++ b/config/opal_check_pmi.m4
@@ -424,9 +424,11 @@ AC_DEFUN([OPAL_CHECK_PMIX],[
                        [Whether the external PMIx library is v1])
     AM_CONDITIONAL([OPAL_WANT_PRUN], [test "$opal_prun_happy" = "yes"])
 
-    AS_IF([test "$opal_external_pmix_version" = "1x"],
-          [OPAL_SUMMARY_ADD([[Miscellaneous]],[[PMIx support]], [opal_pmix], [1.2.x: WARNING - DYNAMIC OPS NOT SUPPORTED])],
-          [OPAL_SUMMARY_ADD([[Miscellaneous]],[[PMIx support]], [opal_pmix], [$opal_external_pmix_version])])
+    AS_IF([test "$opal_external_pmix_happy" = "yes"],
+          [AS_IF([test "$opal_external_pmix_version" = "1x"],
+                 [OPAL_SUMMARY_ADD([[Miscellaneous]],[[PMIx support]], [opal_pmix], [External (1.2.5) WARNING - DYNAMIC OPS NOT SUPPORTED])],
+                 [OPAL_SUMMARY_ADD([[Miscellaneous]],[[PMIx support]], [opal_pmix], [External ($opal_external_pmix_version)])])],
+          [OPAL_SUMMARY_ADD([[Miscellaneous]], [[PMIx support]], [opal_pmix], [Internal])])
 
     OPAL_VAR_SCOPE_POP
 ])

--- a/config/opal_config_subdir_args.m4
+++ b/config/opal_config_subdir_args.m4
@@ -10,7 +10,7 @@ dnl Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
 dnl                         University of Stuttgart.  All rights reserved.
 dnl Copyright (c) 2004-2005 The Regents of the University of California.
 dnl                         All rights reserved.
-dnl Copyright (c) 2014-2015 Intel, Inc. All rights reserved.
+dnl Copyright (c) 2014-2018 Intel, Inc.  All rights reserved.
 dnl Copyright (c) 2015      Cisco Systems, Inc.  All rights reserved.
 dnl $COPYRIGHT$
 dnl
@@ -60,6 +60,8 @@ do
 	    ;;
 	-with-platform=* | --with-platform=*)
 	    ;;
+        -with*=internal)
+            ;;
 	*)
 	    case $subdir_arg in
 	    *\'*) subdir_arg=`echo "$subdir_arg" | sed "s/'/'\\\\\\\\''/g"` ;;

--- a/config/opal_config_subdir_args.m4
+++ b/config/opal_config_subdir_args.m4
@@ -60,7 +60,7 @@ do
 	    ;;
 	-with-platform=* | --with-platform=*)
 	    ;;
-        -with*=internal)
+        --with*=internal)
             ;;
 	*)
 	    case $subdir_arg in

--- a/config/opal_config_subdir_args.m4
+++ b/config/opal_config_subdir_args.m4
@@ -62,6 +62,8 @@ do
 	    ;;
         --with*=internal)
             ;;
+        --with*=external)
+            ;;
 	*)
 	    case $subdir_arg in
 	    *\'*) subdir_arg=`echo "$subdir_arg" | sed "s/'/'\\\\\\\\''/g"` ;;

--- a/opal/mca/pmix/ext2x/Makefile.am
+++ b/opal/mca/pmix/ext2x/Makefile.am
@@ -38,7 +38,7 @@ endif
 
 mcacomponentdir = $(opallibdir)
 mcacomponent_LTLIBRARIES = $(component_install)
-mca_pmix_ext2x_la_SOURCES = $(sources)
+mca_pmix_ext2x_la_SOURCES = $(sources) $(headers)
 mca_pmix_ext2x_la_CFLAGS = $(opal_pmix_ext2x_CFLAGS)
 mca_pmix_ext2x_la_CPPFLAGS = $(opal_pmix_ext2x_CPPFLAGS)
 mca_pmix_ext2x_la_LDFLAGS = -module -avoid-version $(opal_pmix_ext2x_LDFLAGS)
@@ -46,7 +46,7 @@ mca_pmix_ext2x_la_LIBADD = $(top_builddir)/opal/lib@OPAL_LIB_PREFIX@open-pal.la 
 	$(opal_pmix_ext2x_LIBS)
 
 noinst_LTLIBRARIES = $(component_noinst)
-libmca_pmix_ext2x_la_SOURCES = $(sources)
+libmca_pmix_ext2x_la_SOURCES = $(sources) $(headers)
 libmca_pmix_ext2x_la_CFLAGS = $(opal_pmix_ext2x_CFLAGS)
 libmca_pmix_ext2x_la_CPPFLAGS = $(opal_pmix_ext2x_CPPFLAGS)
 libmca_pmix_ext2x_la_LDFLAGS = -module -avoid-version $(opal_pmix_ext2x_LDFLAGS)

--- a/opal/mca/pmix/ext2x/Makefile.am
+++ b/opal/mca/pmix/ext2x/Makefile.am
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2014-2017 Intel, Inc. All rights reserved.
+# Copyright (c) 2014-2018 Intel, Inc.  All rights reserved.
 # Copyright (c) 2015      Cisco Systems, Inc.  All rights reserved.
 # Copyright (c) 2015      Research Organization for Information Science
 #                         and Technology (RIST). All rights reserved.
@@ -40,13 +40,13 @@ mcacomponentdir = $(opallibdir)
 mcacomponent_LTLIBRARIES = $(component_install)
 mca_pmix_ext2x_la_SOURCES = $(sources)
 mca_pmix_ext2x_la_CFLAGS = $(opal_pmix_ext2x_CFLAGS)
-mca_pmix_ext2x_la_CPPFLAGS =$(opal_pmix_ext2x_CPPFLAGS)
+mca_pmix_ext2x_la_CPPFLAGS = $(opal_pmix_ext2x_CPPFLAGS)
 mca_pmix_ext2x_la_LDFLAGS = -module -avoid-version $(opal_pmix_ext2x_LDFLAGS)
 mca_pmix_ext2x_la_LIBADD = $(top_builddir)/opal/lib@OPAL_LIB_PREFIX@open-pal.la \
 	$(opal_pmix_ext2x_LIBS)
 
 noinst_LTLIBRARIES = $(component_noinst)
-libmca_pmix_ext2x_la_SOURCES =$(sources)
+libmca_pmix_ext2x_la_SOURCES = $(sources)
 libmca_pmix_ext2x_la_CFLAGS = $(opal_pmix_ext2x_CFLAGS)
 libmca_pmix_ext2x_la_CPPFLAGS = $(opal_pmix_ext2x_CPPFLAGS)
 libmca_pmix_ext2x_la_LDFLAGS = -module -avoid-version $(opal_pmix_ext2x_LDFLAGS)

--- a/opal/mca/pmix/ext2x/ext2x.h
+++ b/opal/mca/pmix/ext2x/ext2x.h
@@ -37,12 +37,6 @@
 
 BEGIN_C_DECLS
 
-#ifdef OPAL_C_HAVE_VISIBILITY
-#define PMIX_HAVE_VISIBILITY 1
-#else
-#undef PMIX_HAVE_VISIBILITY
-#endif
-
 typedef struct {
   opal_pmix_base_component_t super;
   bool legacy_get;

--- a/opal/mca/pmix/pmix3x/configure.m4
+++ b/opal/mca/pmix/pmix3x/configure.m4
@@ -87,6 +87,9 @@ AC_DEFUN([MCA_opal_pmix_pmix3x_CONFIG],[
           [AC_MSG_RESULT([no - disqualifying this component])
            opal_pmix_pmix3x_happy=0],
           [AC_MSG_RESULT([yes - using the internal v3.x library])
+           AS_IF([test "$opal_pmix_pmix3x_happy" = "0"],
+                 [AC_MSG_WARN([INTERNAL PMIX FAILED TO CONFIGURE])
+                  AC_MSG_ERROR([CANNOT CONTINUE])])
            # Build flags for our Makefile.am
            opal_pmix_pmix3x_LDFLAGS=
            opal_pmix_pmix3x_LIBS="$OPAL_TOP_BUILDDIR/$opal_pmix_pmix3x_basedir/pmix/src/libpmix.la"

--- a/opal/mca/pmix/pmix3x/pmix/config/pmix_setup_hwloc.m4
+++ b/opal/mca/pmix/pmix3x/pmix/config/pmix_setup_hwloc.m4
@@ -24,6 +24,8 @@ AC_DEFUN([PMIX_HWLOC_CONFIG],[
                                 [Search for hwloc libraries in DIR ])])
 
     pmix_hwloc_support=0
+    AS_IF([test "$with_hwloc" = "internal" || test "$with_hwloc" = "external"],
+          [with_hwloc=])
 
     if test "$with_hwloc" != "no"; then
         AC_MSG_CHECKING([for hwloc in])

